### PR TITLE
feat: implement export page

### PR DESCRIPTION
**Pull request by Thposbsmt**
**Youtrack**: https://youtrack.raccoongang.com/issue/2U-161

- Implemented new page - Export page
- Created new generic component - CourseStepper (also will be used in import page)

### CODE REVIEW

File: src/export-page/data/thunks.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  updateSuccessDate,
+  updateError,
+  updateIsErrorModalOpen,
+  reset,
+  updateLoadingStatus,
+} from './slice';
+
+export function startExportingCourse(courseId) {
+  return async (dispatch) => {
+    dispatch(updateLoadingStatus({ status: RequestStatus.IN_PROGRESS }));
```
[the remark this answers is not in the document]
Comment (ruzniaievdm) [reply src/export-page/data/thunks.js:25]: and create slice for `savingStatus`

File: src/export-page/data/slice.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { RequestStatus } from '../../data/constants';
+
+const initialState = {
+  exportTriggered: false,
+  currentStage: 0,
+  error: { msg: null, unitUrl: null },
+  downloadPath: null,
+  successDate: null,
+  isErrorModalOpen: false,
+  loadingStatus: RequestStatus.PENDING,
```
Comment (ruzniaievdm) [inline src/export-page/data/slice.js:13]: please compare the behavior with the help tokens as we discussed

File: src/export-page/data/thunks.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import Cookies from 'universal-cookie';
+
```
Comment (ruzniaievdm) [inline src/export-page/data/thunks.js:2]: this empty line is unnecessary

File: src/export-page/export-footer/ExportFooter.scss [hunk 4]
```diff
@@ -0,0 +1,14 @@
+.export-footer-list {
+  list-style: none;
+  padding-left: 0;
+
+  li {
+    padding-bottom: 5px;
+    border-bottom: 1px solid #E5E5E5;
```
Comment (ruzniaievdm) [inline src/export-page/export-footer/ExportFooter.scss:7]: create please variable for `src/assets/scss/_variables.scss`
with from `src/studio-header/header.scss` too

File: src/export-page/export-footer/ExportFooter.scss [hunk 5]
```diff
@@ -0,0 +1,14 @@
+.export-footer-list {
+  list-style: none;
+  padding-left: 0;
+
+  li {
+    padding-bottom: 5px;
```
Comment (ruzniaievdm) [inline src/export-page/export-footer/ExportFooter.scss:6]: please use `rem`
except `1px`

File: src/export-page/export-stepper/ExportStepper.jsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  injectIntl,
+  intlShape,
+} from '@edx/frontend-platform/i18n';
+import PropTypes from 'prop-types';
+import { useDispatch, useSelector } from 'react-redux';
+import { getConfig } from '@edx/frontend-platform';
+import { Button } from '@edx/paragon';
+
+import CourseStepper from '../../generic/course-stepper';
+
```
Comment (ruzniaievdm) [inline src/export-page/export-stepper/ExportStepper.jsx:12]: please optimize import ordering and remove empty line here

File: src/export-page/export-stepper/ExportStepper.jsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  const currentStage = useSelector(getCurrentStage);
+  const downloadPath = useSelector(getDownloadPath);
+  const successDate = useSelector(getSuccessDate);
+  const loadingStatus = useSelector(getLoadingStatus);
+  const { msg: errorMessage } = useSelector(getError);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    const id = setInterval(() => {
+      if (currentStage === EXPORT_STAGES.SUCCESS || loadingStatus === RequestStatus.FAILED || errorMessage) {
```
Comment (ruzniaievdm) [inline src/export-page/export-stepper/ExportStepper.jsx:32]: can you move this condition in to variable

File: src/export-page/export-stepper/ExportStepper.jsx [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      <h3 className="mt-4">{intl.formatMessage(messages.stepperHeaderTitle)}</h3>
+      <hr />
+      <CourseStepper
+        courseId={courseId}
+        steps={steps}
+        activeKey={currentStage}
+        errorMessage={errorMessage}
+        hasError={!!errorMessage}
+      />
+      {downloadPath && currentStage === EXPORT_STAGES.SUCCESS && <Button href={`${getConfig().STUDIO_BASE_URL}${downloadPath}`}>{intl.formatMessage(messages.downloadCourseButtonTitle)}</Button>}
```
Comment (ruzniaievdm) [inline src/export-page/export-stepper/ExportStepper.jsx:77]: optional: const downloadCourseUrl = new URL(downloadPath, getConfig().STUDIO_BASE_URL).href;

File: src/export-page/utils.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import Cookies from 'universal-cookie';
+import moment from 'moment';
+
+import { LAST_EXPORT_COOKIE_NAME } from './data/constants';
+
+// eslint-disable-next-line import/prefer-default-export
```
Comment (ruzniaievdm) [inline src/export-page/utils.js:6]: do we need to use disable lint here?

File: src/export-page/utils.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+export const setExportCookie = (date, completed) => {
+  const cookies = new Cookies();
+  cookies.set(LAST_EXPORT_COOKIE_NAME, { date, completed }, { path: window.location.pathname });
+};
+
+export const getFormattedSuccessDate = (unixDate) => {
+  const formattedDate = moment(unixDate).utc().format('MM/DD/YYYY');
```
Comment (ruzniaievdm) [inline src/export-page/utils.js:13]: please use defined `DATE_FORMAT`

File: src/export-page/utils.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  const formattedTime = moment(unixDate).utc().format('HH:mm');
```
Comment (ruzniaievdm) [inline src/export-page/utils.js:14]: please use defined `TIME_FORMAT`

File: src/generic/course-stepper/index.jsx [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+      }
+      if (isLastStepDone) {
+        return SuccessDoneIcon;
+      }
+
+      return SettingsIcon;
+    };
+
+    return {
+      StepIcon: getStepIcon(index),
```
Comment (ruzniaievdm) [inline src/generic/course-stepper/index.jsx:42]: please change to `stepIcon`

**Review by ruzniaievdm — CHANGES_REQUESTED** [review]

**Comment by ruzniaievdm** [comment]
@Thposbsmt  during the opening of the PR to upstream please remove `ENABLE_NEW_EXPORT_PAGE` from env files

**Comment by ruzniaievdm** [comment]
@Thposbsmt it seems like we have validation error with duplicate messages

File: src/export-page/data/thunks.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+import moment from 'moment';
+
+import { setExportCookie } from '../utils';
+import { RequestStatus } from '../../data/constants';
+import { EXPORT_STAGES, LAST_EXPORT_COOKIE_NAME } from './constants';
```
Comment (PKulkoRaccoonGang) [inline src/export-page/data/thunks.js:7]:
```suggestion
import { RequestStatus } from '../../data/constants';
import { setExportCookie } from '../utils';
import { EXPORT_STAGES, LAST_EXPORT_COOKIE_NAME } from './constants';
```

File: src/export-page/export-footer/ExportFooter.jsx [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import {
+  injectIntl,
+  intlShape,
+} from '@edx/frontend-platform/i18n';
+import { Layout } from '@edx/paragon';
+
+import messages from './messages';
+
+const ExportFooter = ({ intl }) => (
+  <div className="mt-4">
```
Comment (PKulkoRaccoonGang) [inline src/export-page/export-footer/ExportFooter.jsx:11]: if it's a footer should there be an appropriate tag `<footer>` here?

File: src/export-page/export-footer/ExportFooter.scss (hunk 4 above)
Comment (PKulkoRaccoonGang) [reply src/export-page/export-footer/ExportFooter.scss:7]: Do we have the opportunity to use `$gray-100` [Paragon SCSS variable](https://paragon-openedx.netlify.app/foundations/colors)? Or maybe darken($white, desired percentage shading). It seems to me that after that we will not need to take out this color in a separate variable.

File: src/export-page/export-sidebar/ExportSidebar.jsx [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+import {
+  injectIntl,
+  intlShape,
+} from '@edx/frontend-platform/i18n';
+import PropTypes from 'prop-types';
+import { Button } from '@edx/paragon';
+
+import HelpSidebar from '../../generic/help-sidebar';
+import messages from './messages';
+import { useHelpUrls } from '../../help-urls/hooks';
```
Comment (PKulkoRaccoonGang) [inline src/export-page/export-sidebar/ExportSidebar.jsx:11]:
```suggestion
import HelpSidebar from '../../generic/help-sidebar';
import { useHelpUrls } from '../../help-urls/hooks';
import messages from './messages';
```

File: src/export-page/export-stepper/ExportStepper.scss [hunk 16]
```diff
@@ -0,0 +1,3 @@
+.pgn__stepper-header-step-list {
+  flex-direction: column;
```
Comment (PKulkoRaccoonGang) [inline src/export-page/export-stepper/ExportStepper.scss:2]: Try to use [bootstrap utils](https://getbootstrap.com/docs/4.0/utilities/flex/#direction) for flex-direction declaration

File: src/export-page/utils.js [hunk 17]
```diff
@@ -0,0 +1,16 @@
+import Cookies from 'universal-cookie';
+import moment from 'moment';
+
+import { LAST_EXPORT_COOKIE_NAME } from './data/constants';
+
+// eslint-disable-next-line import/prefer-default-export
+export const setExportCookie = (date, completed) => {
```
Comment (PKulkoRaccoonGang) [inline src/export-page/utils.js:7]: For all functions in the utilities, we decided to add special [JSDocs](https://jsdoc.app/about-getting-started.html) documentation. For [example](https://github.com/raccoongang/frontend-app-course-authoring/blob/2u-master/src/advanced-settings/utils.js)

File: src/generic/course-stepper/CouseStepper.scss [hunk 18]
```diff
@@ -0,0 +1,81 @@
+@keyframes spin {
+  from {
+    transform: rotate(0deg);
+  }
+
+  to {
+    transform: rotate(360deg);
+  }
+}
```
Comment (PKulkoRaccoonGang) [inline src/generic/course-stepper/CouseStepper.scss:9]: We already have an [animation to rotate](https://github.com/raccoongang/frontend-app-course-authoring/blob/2u-master/src/assets/scss/_animations.scss) the element. I suggest using it 🥇

**Review by PKulkoRaccoonGang — COMMENTED** [review]
@Thposbsmt please add a short description of this pull request (what was done) and a link to the YouTrack ticket. Add also a link in YouTrack to this PR.

File: src/export-page/CourseExportPage.jsx [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+            </Layout.Element>
+          </Layout>
+        </section>
+        <ExportModalError courseId={courseId} />
+      </Container>
+      <div className="alert-toast">
+        <InternetConnectionAlert
+          isFailed={loadingStatus === RequestStatus.FAILED}
+          isQueryPending={loadingStatus === RequestStatus.IN_PROGRESS}
+          onInternetConnectionFailed={() => null}
```
[the remark this answers is not in the document]
Comment (Thposbsmt) [reply src/export-page/CourseExportPage.jsx:98]: I don't think so, `onInternetConnectionFailed` marked as `isRequired` in props and haven't default

File: src/export-page/export-stepper/ExportStepper.jsx (hunk 8 above)
Comment (Thposbsmt) [reply src/export-page/export-stepper/ExportStepper.jsx:79]: We dont need it. Backend send full url to studio

**Review by ruzniaievdm — APPROVED** [review]

**Pull request merged**